scipy/sparse/data.py

```diff
@@ -185,13 +187,17 @@ def _min_or_max_axis(self, axis, min_or_max):
             return coo_matrix((value, (major_index, np.zeros(len(value)))),
                               dtype=self.dtype, shape=(M, 1))
 
-    def _min_or_max(self, axis, out, min_or_max):
+    def _min_or_max(self, axis, out, min_or_max, nonzero):
         if out is not None:
             raise ValueError(("Sparse matrices do not support "
                               "an 'out' parameter."))
 
         validateaxis(axis)
 
+        if hasattr(self, 'has_canonical_format'):
+            if not self.has_canonical_format and nonzero:
+                self.eliminate_zeros()
```

Comment:
I don't think we should be eliminating zeros here. There are use-cases that distinguish between explicit and implicit zeros (like the csgraph module), so we can't always assume that explicit zeros should be ignored. It's also an in-place update, which I wouldn't expect to happen when calling `.max()`.

Reply:
I introduced this mainly to have compatibility with the non-canonical forms. There could be a couple of different alternative implementations. The first is it could consider explicit zeros that are left in the matrix. The second could be a forced zero elimination if any values are zeros in the data. The main issue with the former is that it could give unexpected behavior if the user didn't call eliminate zeros prior to calling the function. That may be the best alternative especially if the flag is updated from 'nonzero' to 'explicit' along with the documentation.

Reply:
+1 for changing the name rather than eliminating zeros.

Reply:
The new keyword should be in place. I used "explicit", since it's considering values that are within the actual matrix.

Reply:
I should also mention that I reworked the tests to allow the non-canonical sparse matrices to be correctly validated in the 'explicit' case.